src/listen/ldap_sync/proto_ldap_sync_ldap.c

```diff
@@ -0,0 +1,1325 @@
+
+				cookie = talloc_memdup(inst, fr_dbuff_current(&dbuff), len);
+			}
+		}
+
+		inst->sync_config[packet_id]->init(thread->conn->h, inst->sync_config[packet_id], cookie);
+	}
+		break;
+
+	case FR_LDAP_SYNC_CODE_ENTRY_RESPONSE:
+	{
+		sync_entry_packet_t	*sync_packet;
+
+		if (!packet_ctx) break;
+
+		/*
+		 *	Now the packet has been handled, free the ctx.
+		 *	This will free the LDAPMessage.
+		 */
+		sync_packet = talloc_get_type_abort(packet_ctx, sync_entry_packet_t);
+
+		talloc_free(sync_packet);
+	}
+		break;
+
+	case FR_LDAP_SYNC_CODE_COOKIE_STORE_RESPONSE:
+	{
+		sync_refresh_packet_t	*refresh_packet;
+		sync_config_t const	*sync_config;
+
+		if (!packet_ctx) break;
+
+		/*
+		 *	If there is a packet_ctx, it will be the tracking structure
+		 *	indicating that we need to refresh the sync.
+		 */
+		refresh_packet = talloc_get_type_abort(packet_ctx, sync_refresh_packet_t);
+
+		/*
+		 *	Abandon the old sync and start a new one with the relevant cookie.
+		 */
+		sync_config = refresh_packet->sync->config;
+		DEBUG3("Restarting sync with base %s", sync_config->base_dn);
+		talloc_free(refresh_packet->sync);
+		inst->sync_config[packet_id]->init(thread->conn->h, sync_config, refresh_packet->refresh_cookie);
+
+		talloc_free(refresh_packet);
+	}
+		break;
+
+	default:
+		ERROR("Invalid packet type returned %d", pcode);
+		break;
+	}
+
+	return buffer_len;
+}
+
+/** Callback for socket errors when running initial root query
+ */
+static void _proto_ldap_socket_open_error(UNUSED fr_event_list_t *el, UNUSED int fd, UNUSED int flags,
+					  UNUSED int fd_errno, void *uctx)
+{
+	proto_ldap_dir_ctx	*dir_ctx = talloc_get_type_abort(uctx, proto_ldap_dir_ctx);
+	fr_ldap_connection_t	*ldap_conn = talloc_get_type_abort(dir_ctx->conn->h, fr_ldap_connection_t);
+
+	talloc_free(dir_ctx);
+	fr_ldap_state_error(ldap_conn);
+}
+
+/** Callback to process results of initial root query, identifying directory type
+ */
+static void _proto_ldap_socket_open_read(fr_event_list_t *el, int fd, UNUSED int flags, void *uctx)
+{
+	proto_ldap_dir_ctx		*dir_ctx = talloc_get_type_abort(uctx, proto_ldap_dir_ctx);
+	fr_ldap_connection_t		*ldap_conn = talloc_get_type_abort(dir_ctx->conn->h, fr_ldap_connection_t);
+	proto_ldap_sync_ldap_t const	*inst = talloc_get_type_abort_const(dir_ctx->main_listen->app_io_instance,
+									    proto_ldap_sync_ldap_t);
+	proto_ldap_sync_ldap_thread_t	*thread = talloc_get_type_abort(dir_ctx->main_listen->thread_instance,
+									proto_ldap_sync_ldap_thread_t);
+	fr_ldap_rcode_t			status;
+	LDAPMessage			*result;
+
+	size_t				i;
+
+	/*
+	 *	Timeout has to be > 0 msec, otherwise libldap always returns LDAP_TIMEOUT
+	 *	Very short timeouts have also been seen to return LDAP_TIMEOUT even though
+	 *	there is data to read, depending on computer speed.  10ms appears to be safe.
+	 */
+	status = fr_ldap_result(&result, NULL, ldap_conn, dir_ctx->msgid, LDAP_MSG_ALL, NULL, fr_time_delta_from_msec(10));
```

Comment:
I don't really understand what's happening here.  Why is there a fixed timeout?  Won't open_Read be called on an I/O event?

Reply:
I think when this was originally coded I didn't have `res_timeout` being set correctly.  If the timeout that `ldap_result()` is called with is too short then it returns a TIMEOUT even though there is data available - maybe due to the time it is taking to parse what's come in on the wire.
Having got `res_timeout` plumbed in correctly, we can call things with zero there and it's reliable.